### PR TITLE
LWM2M: Tests covering new functionality from GSoC

### PR DESCRIPTION
**Pull request by bsmelo**
This patch adds lwm2m tests covering all new functionality
added during GSoC, including: Bootstrap Interface, Access
Control Objects and CoAP Encryption (DTLS Security Modes).

Depends on #2290.

Signed-off-by: Bruno Melo <bsilva.melo@gmail.com>

### CODE REVIEW

File: src/test/test-lwm2m.c [hunk 1]
```diff
@@ -460,7 +1169,7 @@ observe_res_cb(void *data,
     enum sol_lwm2m_content_type content_type,
     struct sol_str_slice content)
 {
-    static int i = 0;
+    static int nosec_i = 0, sec_i = 0;
```
Comment (cabelitos) [inline src/test/test-lwm2m.c]: I would create an ENUM to describe which state we're right know.

File: src/test/test-lwm2m.c [hunk 2]
```diff
@@ -562,94 +1322,522 @@ delete_cb(void *data,
...
+#ifdef DTLS
+    struct sol_lwm2m_bootstrap_server *bs_server;
+    struct sol_lwm2m_server *sec_server;
+    struct sol_lwm2m_client *sec_client;
+    static const struct sol_lwm2m_object *sec_objects[] =
+    { &security_object, &server_object,
+      &access_control_object, &dummy_object, NULL };
+    struct security_obj_instance_ctx *sec_security_data;
+    struct sol_lwm2m_security_psk *bs_psk, *sec_psk;
+    struct sol_vector bs_known_keys = { }, sec_known_keys = { };
```
Comment (cabelitos) [inline src/test/test-lwm2m.c]: you can use
`struct sol_vector bs_known_keys = SOL_VECTOR_INIT(struct sol_lwm2m_security_psk);`

File: src/test/test-lwm2m.c [hunk 3]
```diff
@@ -562,94 +1322,522 @@ delete_cb(void *data,
...
+    struct sol_lwm2m_bootstrap_server *server,
+    struct sol_lwm2m_bootstrap_client_info *bs_cinfo)
+{
+    int r;
+
+    ASSERT(!strcmp("cli1", sol_lwm2m_bootstrap_client_info_get_name(bs_cinfo)));
+
+    printf("DBG: [Bootstrap Delete /]---> '%s'\n",
+        sol_lwm2m_bootstrap_client_info_get_name(bs_cinfo));
+    r = sol_lwm2m_bootstrap_server_delete_object_instance(server, bs_cinfo, "/",
```
Comment (cabelitos) [inline src/test/test-lwm2m.c]: I guess we should add a test that a server will send a delete on '/' while NOT in bootstrap.

Comment (cabelitos) [inline src/test/test-lwm2m.c]: To make sure that we implemented correctly

File: src/test/test-lwm2m.c [hunk 4]
```diff
@@ -562,94 +1322,522 @@ delete_cb(void *data,
...
+static void
+read_unauthorized_cb(void *data,
+    struct sol_lwm2m_server *server,
+    struct sol_lwm2m_client_info *client,
+    const char *path,
+    enum sol_coap_response_code response_code,
+    enum sol_lwm2m_content_type content_type,
+    struct sol_str_slice content)
+{
+    ASSERT(response_code == SOL_COAP_RESPONSE_CODE_UNAUTHORIZED);
```
Comment (cabelitos) [inline src/test/test-lwm2m.c]: I think we should chain the events and not leave them hanging around like this.

File: src/test/test-lwm2m.c [hunk 5]
```diff
@@ -562,94 +1322,522 @@ delete_cb(void *data,
...
+            SOL_LWM2M_RESOURCE_DATA_TYPE_BOOL, false);
+        ASSERT(r == 0);
+        SOL_LWM2M_RESOURCE_SINGLE_INIT(r, &res[1], DUMMY_OBJECT_FLOAT_ID,
+            SOL_LWM2M_RESOURCE_DATA_TYPE_FLOAT, FLOAT_VALUE);
+        ASSERT(r == 0);
+
+        printf("DBG: '%s' ---[Write /999]---> '%s'\n",
+            server_type, sol_lwm2m_client_info_get_name(cinfo));
+        r = sol_lwm2m_server_write(server, cinfo, "/999",
+            res, sol_util_array_size(res), NULL, data);
```
Comment (cabelitos) [inline src/test/test-lwm2m.c]: Why the callback here is NULL ?

File: src/test/test-lwm2m.c (hunk 3 above)
Comment (cabelitos) [inline src/test/test-lwm2m.c]: Nevermind. You're already doing this.

**Comment by cabelitos** [comment]
It would be nice to add comments about what kind of objects each client has and what the servers are trying to do. Could you add this documentation? You should also add to this docs operations that the servers will try to do, but will fail since they are not allowed.

File: src/test/test-lwm2m.c (hunk 1 above)
Comment (bsmelo) [inline src/test/test-lwm2m.c]: Which states would those be? I must confess I didn't "deep dive" in this part - the sequence of tests performed originally, which apparently includes a step of modifying one integer resource and so on. I just changed the `static i` to `static "i(s)"` to support two different clients independently.

File: src/test/test-lwm2m.c (hunk 4 above)
Comment (bsmelo) [inline src/test/test-lwm2m.c]: Why? I mean, I'm only doing this "hang around" thing for the negative- and corner-cases. Since they should fail anyway there's no dependency between the operations (like a Create needing to be executed before a Read can happen and so on).

File: src/test/test-lwm2m.c (hunk 5 above)
Comment (bsmelo) [inline src/test/test-lwm2m.c]: Because it must fail without a CoAP Request being even sent to the Client. Since the callback would never be called anyway (there's no Request -> there's no Response), we save code.

Comment (cabelitos) [inline src/test/test-lwm2m.c]: God. You're right. It's funny, because I fixed this problem. lol

File: src/test/test-lwm2m.c (hunk 1 above)
Comment (cabelitos) [inline src/test/test-lwm2m.c]: States:
INT_VALUE_NOT_SET and INT_VALUE_REPLACE

Comment (cabelitos) [inline src/test/test-lwm2m.c]: I know, I just want to keep it more readable.

**Comment by bsmelo** [comment]
An updated version is at #2294.
